test/tempfile.c

```diff
@@ -0,0 +1,104 @@
+/*
+ * Schism Tracker - a cross-platform Impulse Tracker clone
+ * copyright (c) 2003-2005 Storlek <storlek@rigelseven.com>
+ * copyright (c) 2005-2008 Mrs. Brisby <mrs.brisby@nimh.org>
+ * copyright (c) 2009 Storlek & Mrs. Brisby
+ * copyright (c) 2010-2012 Storlek
+ * URL: http://schismtracker.org/
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#include "headers.h"
+#include "util.h"
+
+#include "test-tempfile.h"
+
+#ifdef SCHISM_WIN32
+#define unlink _unlink
+#endif
+
+int test_temp_file(char **temp_file, const char *template, int template_length)
+{
+	*temp_file = malloc(16);
+
+	strcpy(*temp_file, "test_tmp_XXXXXX");
+
+	FILE *f = mkfstemp(*temp_file);
+
+	if (f == NULL)
+	{
+		free(*temp_file);
+		*temp_file = NULL;
+		return 0;
+	}
+
+	int remaining = 0;
+
+	if (template != NULL) {
+		if (template_length < 0)
+			template_length = strlen(template);
+
+		const char *buf = template;
+		remaining = template_length;
+
+		while (1) {
+			int num_written = fwrite(buf, 1, remaining, f);
```

Comment:
i think `fwrite` can only return less than `remaining` if an error occurs (in which case any further writes will be blocked)

Reply:
According the documentation for `fwrite`, there are plenty of transitive reasons why a write might be partial:

* Insufficient disk space (could change at any time)
* Interrupted syscalls
* Device-specific limitations ("Some devices or file systems might impose limitations that lead to partial writes in specific scenarios")

When a partial write happens, `fwrite` doesn't return an error. I'm pretty sure it is canonically correct to treat a partial write as a success and retry with the remaining data. If the error was not transient, then the next call will fail to write any items. But, the way the interface is defined, the next call _could_ succeed and write more items.